Cleanups to port mapping module post UPnP drop

darosior opened the pull request:
Followup to #31130, this does a couple cleanups to `src/mapport.*` to clarify the logic now that there is a single protocol option for port mapping.

darosior committed:
mapport: make 'enabled' and 'current' bool

Since there is only a single protocol now, clarify the code by changing
the protocol enum for a bool for both variables.

laanwj commented:
i'll review this when #31130 is merged, i have some of my own ideas about portmap.cpp as well that i haven't fully worked out yet.

darosior commented:
Rebased after #31130 got merged.

laanwj commented on src/qt/optionsmodel.cpp:
```diff
@@ -526,7 +526,7 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value, const std::
         if (changed()) {
             update(value.toBool());
-            node().mapPort(value.toBool());
+            if (value.toBool()) node().mapPort();
```
By adding this if here, this no longer accounts for the situation where port mapping is turned off. i think i slightly prefer the old API.

darosior replied:
Ah, yeah, right, that was stupid. I'll stick to just renaming `use_pcp` to `enabled`.

darosior replied:
Done.

darosior committed:
mapport: rename 'use_pcp' to 'enable'

There is only a single protocol now, caller should just be concerned about whether to enable port mapping or not.

darosior committed:
mapport: drop unnecessary function

laanwj commented on src/mapport.cpp:
```diff
@@ -153,44 +153,23 @@ void StartThreadMapPort()
 
 static void DispatchMapPort()
```
DispatchMapPort is local and only called in one place; might as well roll it into `StartMapPort`?

darosior replied:
Done.

laanwj commented on src/mapport.cpp:
```diff
@@ -153,44 +153,23 @@ void StartThreadMapPort()
-    }
-
-    if (g_mapport_current_proto == MapPortProtoFlag::NONE && g_mapport_enabled_protos != MapPortProtoFlag::NONE) {
+    if (!g_mapport_current && g_mapport_enabled) {
```
Maybe i'm missing something but the intent of `g_mapport_current` is hazy to me. It's turned on in `ThreadMapPort` while `ProcessPCP` is running, then turned off temporarily when `ProcessPCP` returns false.

- If we want to know if the mapport thread is running, `g_mapport_thread.joinable()` seems to suffice.
- If we want to know if mapport thread is supposed to be running, or stop it, there's `g_mapport_enabled`.

darosior replied:
Yes, as far as i can tell it's an artifact of switching between protocols. I should remove it. In fact i think `g_mapport_enabled` can be removed as well as we can just use `g_mapport_interrupt` to stop it.

darosior replied:
Done for both `current` and `enabled`.

laanwj replied:
Nice! Good catch.

darosior committed:
mapport: merge DispatchMapPort into StartMapPort

darosior committed:
mapport: remove unnecessary 'g_mapport_current' variable

darosior committed:
mapport: remove unnecessary 'g_mapport_enabled'

It was only necessary for switching between mapping protocols. It's also used to return
in ThreadMapPort but we can just use the interrupt for this purpose.

laanwj commented on src/mapport.cpp:
```diff
@@ -125,22 +123,7 @@ static bool ProcessPCP()
-        }
-
-    } while (ok || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD));
+    while ((!g_mapport_interrupt && ProcessPCP()) || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD)) {}
```
The thread is not supposed to exit if `ProcessPCP()` fails. It should be more persistent: it should just wait a while and try again, maybe for the user to join a different network, swap their router, upgrade firmware, etc.

darosior replied:
Yes, this will not exit if  `ProcessPCP` fails. (Unless it's interrupted of course.)

darosior replied:
Maybe i should make this loop clearer? Like:
```diff
diff --git a/src/mapport.cpp b/src/mapport.cpp
index b303017144..8df78f435d 100644
--- a/src/mapport.cpp
+++ b/src/mapport.cpp
@@ -123,7 +123,9 @@ static bool ProcessPCP()
 
 static void ThreadMapPort()
 {
-    while ((!g_mapport_interrupt && ProcessPCP()) || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD)) {}
+    do {
+	ProcessPCP();
+    } while (g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD)) {}
 }
 
 void StartThreadMapPort()
```

Which also makes clearer that `ProcessPCP()` should actually be `void`.

laanwj replied:
Yes, that seems better. Making ProcessPCP a void is also fine with me.

darosior replied:
Done.

darosior committed:
mapport: make ProcessPCP void

Its return value is now unused. (It was also effectively unused before the previous commit, just in a roundabout way).

darosior commented:
I was playing with a couple routers and figured i'd test this. Could successfully map ports against an OPNSense box and a Spectrum "[WIFI 6E router](https://www.spectrum.com/internet/wifi-service/spectrum-advanced-wifi)".

laanwj reviewed (approved):
Code review ACK 70398ae05bc36a2788e87f67ae06962f43fe35a6

TheCharlatan commented on src/mapport.cpp:
```diff
@@ -119,28 +117,13 @@ static bool ProcessPCP()
 
     // We don't delete the mappings when the thread is interrupted because this would add additional complexity, so
     // we rather just choose a fairly short expiry time.
```
This comment seems to be lacking context now.

darosior replied:
Why? If we are interrupted the `while()` clause above will continue here and we won't tell the router to remove the mappings and just return instead.

sedited reviewed (approved):
ACK 70398ae05bc36a2788e87f67ae06962f43fe35a6

Just a nit: Maybe cleanup the dangling includes?
